### PR TITLE
Docks!

### LINKED ISSUE #17: Needs a visual way to close it.
I accidentally opened this thing, had no idea what it was, and no idea how to close it. I had to google "atom keybinding resolver" to find this page. It needs an 'X' button somewhere.

### LINKED ISSUE #24: 'Esc' not resolved
`escape` is a valid shortcut (Atom uses it, one can override it), and well, keybinding-resolver doesn't resolve it - its dialog gets closed instead.

The current behavior is quite debatable - could you give it a think?

Thanks - Victor

### LINKED ISSUE #27: Feature/Behavior
When you're typing at the bottom of the editor pane, this window pops up and hides your current cursor (along with the code you're currently editing).  Rather than toggling the resolver altogether, it'd be nice if there was an option to fix the keyboard-resolver window size at the bottom, or make it scroll the editor window to keep your cursor visible.

### PR DESCRIPTION
**Pull request by 50Wliu**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

Converts the Keybinding Resolver into a Dock.  This fixes some usability issues with the resolver, such as:
* Variable height causing the keybinding resolver to obscure the entire workspace when a large amount of keybindings are present.  Since it's now constrained in a dock, the keybindings will scroll instead.
* Lack of a close button
* Inability to view whatever was bound to `core:cancel` (due to it closing the resolver first)

And adds a nice styling touch to make the header sticky.

### Alternate Designs

None.

### Benefits

See above.

### Possible Drawbacks

I don't foresee any.  This should strictly be an upgrade from the existing panel, though I may be missing something.  One small thing is that after using the resolver, you might be inclined to close the dock instead of the keybinding resolver.  In that case the resolver will still be active and continue to track and display keybindings, resulting in a small performance penalty.

### Applicable Issues

Closes #17
Fixes #24
Closes #27
Supersedes and closes #53

/cc @Ben3eeE @UziTech

Most of the boilerplate code was copied from the active-editor-info example package.  I don't want to needlessly ping someone, but thanks a lot `@matthewwithanm` for writing that :).  Helped me a lot!

### CODE REVIEW

**Comment by 50Wliu** [comment]
Remaining concerns:
* The keybinding resolver continues to receive keybinding events even when hidden.  Is there a way to prevent this behavior to avoid taking a small performance hit?
* `Key Binding` or `Keybinding`?  Lots of things use `keybinding` but there are some places (like the command itself!) that uses `key-binding`.
* Panel styles are still being used despite not being a panel anymore

**Comment by ungb** [comment]
I think this is good, I was able to test it now.  

Although if you hide the dock(as you mention), it will continue to capture.  I think this is acceptable.   

I'm all for merging this, but going to /cc @simurai if he has any feedback on the styling.

screenshot: 
image

File: styles/keybinding-resolver.less [hunk 1]
```diff
@@ -2,14 +2,24 @@
 @import "octicon-mixins";
 
 .key-binding-resolver {
-  -webkit-flex: 0.25;
   overflow: auto;
 
-  .keystroke {
-    margin-right: 1em;
+  .panel-heading {
+    position: sticky;
```
Comment (simurai) [inline styles/keybinding-resolver.less]: Ohh.. wow. Didn't know that works now. We could use that in other places. Like root folders in the tree-view or package titles in the settings.

**Comment by simurai** [comment]
Yay for no more jumping.. 🎉 

Added some small restyling:

Before | After
--- | ---
screen shot 2018-01-26 at 1 45 17 pm | screen shot 2018-01-26 at 1 39 36 pm

The header background could be changed to look seamless to the tab, but that's something themes have to look into.

> The keybinding resolver continues to receive keybinding events even when hidden. Is there a way to prevent this behavior to avoid taking a small performance hit?

Do Docks have some sort of "only run if dock item is visible" option? Maybe [::isVisible()](https://atom.io/docs/api/v1.23.3/Dock#instance-isVisible)? So that it can be paused while the dock item is hidden.

**Comment by ungb** [comment]
LGTM! 🚢

File: styles/keybinding-resolver.less (hunk 1 above)
Comment (50Wliu) [reply styles/keybinding-resolver.less]: :heart_eyes:

**Comment by 50Wliu** [comment]
I'll see if I can wire something up with `onDidChangeVisible`.

**Comment by bdowling** [comment]
fwiw :+1: on this one, I merged it with my PR #62 and it is much more pleasant to use in that mode without it jumping around and covering the editor..

**Comment by 50Wliu** [comment]
I've been quite busy.  Intend to return to this one soon.

**Comment by bdowling** [comment]
fwiw, my merge is here, https://github.com/bdowling/keybinding-resolver/tree/add_copy_keybinding_with_pr58 -- it's been working well for me, other than the minor conflicts, not sure what else this PR needs to get accepted, but I like it.  Thanks!

**Comment by 50Wliu** [comment]
Ok, so the issue is isVisible is for the whole dock, and the Resolver doesn't necessarily have to be in focus for isVisible to return true.

**Comment by 50Wliu** [comment]
Think I fixed all outstanding concerns! Just in time for the one-year anniversary of this PR :)

**Comment by 50Wliu** [comment]
Merging based off of LGTM from @ungb and other positive comments.

**Comment by nathansobo** [comment]
Thanks for this. Sorry it took so long.

**Pull request merged**